### PR TITLE
Add initGame and joinGame to contract. #44 #45

### PR DESCRIPTION
**Pull request by graup**

### CODE REVIEW

**Comment by graup** [comment]
Jenkins, again, please!!

**Comment by Kyroy** [comment]
Jenkins ?

**Comment by Kyroy** [comment]
Jenkins test pls

**Comment by Kyroy** [comment]
Jenkins

**Comment by Kyroy** [comment]
Jenkins

**Comment by Kyroy** [comment]
Jenkins

**Comment by Kyroy** [comment]
Jenkins

**Comment by Kyroy** [comment]
Jenkins

**Comment by Kyroy** [comment]
Jenkins

**Comment by Kyroy** [comment]
Jenkins

**Comment by Kyroy** [comment]
Jenkins

**Comment by Kyroy** [comment]
Jenkins

**Comment by Kyroy** [comment]
Jenkins test please

**Comment by Kyroy** [comment]
Jenkins

**Comment by dennis-westphal** [comment]
This is missing the possibility to define a start color.

**Comment by graup** [comment]
Do we really need that feature? It would be MUCH easier if the initiator is always player 1. 
On 19 May 2016 8:38 pm, dennis-westphal <notifications@github.com> wrote:This is missing the possibility to define a start color.

**Comment by dennis-westphal** [comment]
This is what we discussed and what the story and task say. I aggree that the initiator should always be player 1, but he should not be fixed to playing with white.

**Comment by Kyroy** [comment]
Jenkins

**Comment by Kyroy** [comment]
Jenkins

**Comment by graup** [comment]
OK, so we should have an additional variable that saves who is white. Any suggestion how the function call should look like? e.g. just `true`/`false` for white or not? `Chess.initGame('Alice', true, {from: player1, gas: 1000000});`

**Comment by dennis-westphal** [comment]
Yes, another variable would be good. As for the method signature, normally I would say accepting a String and checking if it is white or black (and throw an error if neither), but if it's easier for you, a boolean isPlayingWhite would be ok too.

**Comment by dennis-westphal** [comment]
Can you please also change the joinGame event so that it also contains the player1 and player1alias? Makes it way easier to know if the game that someone joined belongs to yourself.

File: test/test.js [hunk 1]
```diff
@@ -12,29 +12,56 @@ describe('Chess', function() {
     it('should initialize a game between two players', function (done) {
       this.timeout(10000);
 
-      Chess.initGame(player2, {from: player1, gas: 1000000});
+      Chess.initGame('Alice', {from: player1, gas: 1000000});
 
       var filter = Chess.GameInitialized({});
```
Comment (Pepperrs) [inline test/test.js]: Kommentare wären schön

File: test/test.js [hunk 2]
```diff
@@ -12,29 +12,56 @@ describe('Chess', function() {
...
+        assert.equal('Bob', result.args.player2Alias);
+        filter.stopWatching();
+        done();
+      });
+    });
+    it('should disallow joing a game with two payers', function() {
+      this.timeout(10000);
+
+      assert.throws(function(){
+        Chess.joinGame(gameId, 'Bob', {from: player2, gas: 100000});
```
Comment (Pepperrs) [inline test/test.js]: Ich behaupte der Test geht nicht, da Tests nicht immer in der gleichen Reihenfolge ablaufen (sollten). Sie sollten synchron laufen. 
So wie ich das hier verstehe erwartest du, dass in "Chess" bereits zwei Spieler sind, was nur gegeben ist, wenn das vorherige it("") bereits durchgelaufen ist. (Zeile 33).

File: test/test.js [hunk 3]
```diff
@@ -12,29 +12,56 @@ describe('Chess', function() {
...
-        Chess.move(gameId, {from: player1, gas: 100000});
+        Chess.move(gameId, {from: player2, gas: 100000});
       }, Error);
     });
     it('should accept a valid move', function(done) {
       this.timeout(10000);
 
       assert.doesNotThrow(function(){
-        Chess.move(gameId, {from: player2, gas: 100000});
+        Chess.move(gameId, {from: player1, gas: 100000});
```
Comment (Pepperrs) [inline test/test.js]: wo ist hier der Unterschied zu Zeile 57? wenn noch nicht implementiert könntest du xit anstelle von it benutzen

File: test/test.js (hunk 2 above)
Comment (graup) [inline test/test.js]: Du hast recht, das ist nicht ganz sauber, funktioniert aber, weil alle `it` hintereinander ausgeführt werden. Ich denke das ist an dieser Stelle ok, sonst würde man das doppelt testen. Allgemein hast du recht.

File: test/test.js (hunk 3 above)
Comment (graup) [inline test/test.js]: Habe den Unterschied in einem neuen Kommentar erklärt.

**Comment by graup** [comment]
I ended up adding the option to set who is the white player as a boolean parameter to `initGame`: `bool playAsWhite`. Comparing strings isn't easy in Solidity.

The address of the player who is white is also broadcast in the `GameInitialized` and `GameJoined` event. In the former case it will be `0` if the game was initialized with player1 as black.

**Comment by lisasi** [comment]
According to the FIDE chess rules, the player with the color white starts the game. So your code needs to be modified.

> "Der Spieler mit den hellen Figuren (Weiß) macht den ersten Zug, dann ziehen die Spieler abwechselnd, wobei der Spieler mit den dunklen Figuren (Schwarz) den nächsten Zug ausführt. "

**Comment by graup** [comment]
@lisasi good catch, thank you!

**Comment by Kyroy** [comment]
@lisasi @Pepperrs @dennis-westphal Can this now be merged? I think your comments have been fixed. This pull request should be merged before #56 is merged.

**Pull request merged**